### PR TITLE
[InLong-4141][Sort] Sort lightwieght support load data from Pulsar

### LINKED ISSUE #4141: [Feature][Sort] Sort lightwieght support load data from pulsar
### Description

Sort lightwieght support load data to pulsar 

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by thexiay**
### Title Name: [INLONG-4141][Sort] Sort lightwieght support load data from pulsar

Fixes #4141

### Motivation

Sort lightwieght support load data from pulsar

### Modifications

Add Pulsar Extrace Node

### CODE REVIEW

File: inlong-sort/sort-connectors/pom.xml [hunk 1]
```diff
@@ -113,6 +118,13 @@
             <groupId>org.apache.pulsar</groupId>
             <artifactId>pulsar-client-all</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.streamnative.connectors</groupId>
+            <artifactId>pulsar-flink-connector_2.11</artifactId>
+            <version>1.13.6.1-rc9</version>
```
Comment (healchow) [inline inlong-sort/sort-connectors/pom.xml:124]: Please put the version number into the pom file of the root path, thanks.

Comment (thexiay) [reply inlong-sort/sort-connectors/pom.xml:124]: put it in root pom.xml

File: inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/PulsarSqlParserTest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                                new FieldInfo("name", new StringFormatInfo())));
+        return new KafkaLoadNode("1", "kafka_output", fields, relations, null, null,
+                "workerJson", "localhost:9092",
+                new JsonFormat(), null,
+                null, null);
+    }
+
+    /**
+     * build hbase load node
+     * @return hbase load node
```
Comment (gong) [inline inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/PulsarSqlParserTest.java:69]: `hbase` should be `pulsar`

File: inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/PulsarSqlParserTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PulsarSqlParserTest {
+
+    /**
+     * build mysql extract node
+     * @return Mysql extract node
```
Comment (gong) [inline inlong-sort/sort-single-tenant/src/test/java/org/apache/inlong/sort/singletenant/flink/parser/PulsarSqlParserTest.java:51]: `mysql` should be `kafka`

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            options.put("connector", "pulsar-inlong");
+        } else {
+            options.put("connector", "upsert-pulsar-inlong");
+        }
+        options.put("generic", "true");
+        options.put("service-url", serviceUrl);
+        options.put("admin-url", adminUrl);
+        options.put("topic", topic);
+        options.put("scan.startup.mode", scanStartupMode);
+        options.putAll(format.generateOptions());
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:103]: json/avro format need append prefix `key.` and `value.`  Reference to tableOptions of `KafkaExtractNode`

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+@EqualsAndHashCode(callSuper = true)
+@JsonTypeName("pulsarExtract")
+@Data
+public class PulsarExtractNode extends ExtractNode {
```
Comment (gong) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:41]: `ExtractNode` and `Node` missing @JsonSubTypes.Type(value = PulsarExtractNode.class, name = "pulsarExtract")

Comment (gong) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:41]: Add a UT for serialize and deserialize test in inlong-sort/sort-common/src/test/java/org/apache/inlong/sort/protocol/node/extract

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java (hunk 4 above)
Comment (thexiay) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:103]: format option is ok. 'format.' option will be contained in format.generateOptions()

File: inlong-sort/sort-connectors/src/main/java/org/apache/inlong/sort/flink/pulsar/table/UpsertPulsarDynamicTableFactory.java [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        FactoryUtil.TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+
+        ReadableConfig tableOptions = helper.getOptions();
+        DecodingFormat<DeserializationSchema<RowData>> keyDecodingFormat = helper.discoverDecodingFormat(
+                DeserializationFormatFactory.class,
+                KEY_FORMAT);
+        DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat = helper.discoverDecodingFormat(
+                DeserializationFormatFactory.class,
+                VALUE_FORMAT);
```
Comment (gong) [inline inlong-sort/sort-connectors/src/main/java/org/apache/inlong/sort/flink/pulsar/table/UpsertPulsarDynamicTableFactory.java:115]: Please see this code. The ` tableOptions` function of `PulsarExtractNode` should be use format.generateOptions(true) when connector is `upsert-pulsar-inlong`.

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java (hunk 4 above)
Comment (gong) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:103]: > format option is ok. 'format.' option will be contained in format.generateOptions()

#https://github.com/apache/incubator-inlong/pull/4231#discussion_r874685230

File: pom.xml [hunk 7]
```diff
@@ -977,6 +978,12 @@
                 <version>${flink.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>io.streamnative.connectors</groupId>
+                <artifactId>pulsar-flink-connector_2.11</artifactId>
```
Comment (yunqingmoswu) [inline pom.xml:983]: The scala version of 2.11 can be replaced by variables?

File: inlong-sort/sort-connectors/pom.xml [hunk 8]
```diff
@@ -139,6 +144,12 @@
             <artifactId>hadoop-minicluster</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.streamnative.connectors</groupId>
+            <artifactId>pulsar-flink-connector_2.11</artifactId>
```
Comment (yunqingmoswu) [inline inlong-sort/sort-connectors/pom.xml:149]: The scala version of 2.11 can be replaced by variables?

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            @JsonProperty("properties") Map<String, String> properties,
+
+            @Nonnull @JsonProperty("topic") String topic,
+            @Nonnull @JsonProperty("adminUrl") String adminUrl,
+            @Nonnull @JsonProperty("serviceUrl") String serviceUrl,
+            @Nonnull @JsonProperty("format") Format format,
+            @Nonnull @JsonProperty("scanStartupMode") String scanStartupMode,
+            @JsonProperty("primaryKey") String primaryKey) {
+        super(id, name, fields, watermarkField, properties);
+        this.topic = Preconditions.checkNotNull(topic, "pulsar topic is empty");
```
Comment (yunqingmoswu) [inline inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:77]: Would it be better to keep the validation message and validation logic all the time? Maybe it is "xxx is null"?

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java (hunk 5 above)
Comment (thexiay) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:41]: ok

File: inlong-sort/sort-connectors/src/main/java/org/apache/inlong/sort/flink/pulsar/table/UpsertPulsarDynamicTableFactory.java (hunk 6 above)
Comment (thexiay) [reply inlong-sort/sort-connectors/src/main/java/org/apache/inlong/sort/flink/pulsar/table/UpsertPulsarDynamicTableFactory.java:115]: ok,i will modify it.

File: pom.xml (hunk 7 above)
Comment (thexiay) [reply pom.xml:983]: done

File: inlong-sort/sort-connectors/pom.xml (hunk 8 above)
Comment (thexiay) [reply inlong-sort/sort-connectors/pom.xml:149]: done

File: inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java (hunk 9 above)
Comment (thexiay) [reply inlong-sort/sort-common/src/main/java/org/apache/inlong/sort/protocol/node/extract/PulsarExtractNode.java:77]: done

**Review by yunqingmoswu — APPROVED** [review]

**Review by gong — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

**Pull request merged**